datafusion/src/optimizer/remove_duplicate_filters.rs

```diff
@@ -0,0 +1,611 @@
+        Expr::Literal(ScalarValue::Float32(Some(v))) if *v == 1. => true,
+        Expr::Literal(ScalarValue::Float64(Some(v))) if *v == 1. => true,
+        _ => false
+    }
+}
+
+fn is_zero<'a>(s: &'a Expr) -> bool {
+    match s {
+        Expr::Literal(ScalarValue::Int8(Some(0))) => true,
+        Expr::Literal(ScalarValue::Int16(Some(0))) => true,
+        Expr::Literal(ScalarValue::Int32(Some(0))) => true,
+        Expr::Literal(ScalarValue::Int64(Some(0))) => true,
+        Expr::Literal(ScalarValue::UInt8(Some(0))) => true,
+        Expr::Literal(ScalarValue::UInt16(Some(0))) => true,
+        Expr::Literal(ScalarValue::UInt32(Some(0))) => true,
+        Expr::Literal(ScalarValue::UInt64(Some(0))) => true,
+        Expr::Literal(ScalarValue::Float32(Some(v))) if *v == 0. => true,
+        Expr::Literal(ScalarValue::Float64(Some(v))) if *v == 0. => true,
+        _ => false
+    }
+}
+
+fn is_true<'a>(expr: &'a Expr) -> bool {
+    match expr {
+        Expr::Literal(ScalarValue::Boolean(Some(v))) => *v,
+        _ => false,
+    }
+}
+
+fn is_false<'a>(expr: &'a Expr) -> bool {
+    match expr {
+        Expr::Literal(ScalarValue::Boolean(Some(v))) => *v == false,
+        _ => false,
+    }
+}
+
+fn simplify<'a>(expr: &'a Expr) -> Expr {
+    match expr {
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if is_true(left) || is_true(right) => lit(true),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if is_false(left) => simplify(right),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if is_false(right) => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if left == right => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if is_false(left) || is_false(right) => lit(false),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if is_true(right) => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if is_true(left) => simplify(right),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if left == right => simplify(right),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Minus,
+            right
+        } if is_zero(left) => Expr::Negative(Box::new(simplify(right))),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Minus,
+            right
+        } if is_zero(right) => simplify(left),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Minus,
+            right
+        } if left == right => lit(0),
```

Comment:
I removed those by now from this PR https://github.com/apache/arrow-datafusion/pull/441 as I think they might be wrong in precense of nulls (null - null` is not `0` ?) `null * 0` is not `0`

Reply:
I agree, already removed.
maybe there should be some specific optimizations for arithmetic operators ...

Reply:
It might be possible to do the optimization if we know the input is non null based on the schema. But probably something to find out later 👍